content/articles/sharing-deployments-with-nixops.md

```diff
@@ -0,0 +1,198 @@
+
+```bash
+export AWS_ACCESS_KEY_ID="AKIA...."
+export AWS_SECRET_ACCESS_KEY="eexie9Ieeel9XieCThish5..."
+export NIX_PATH=${NIX_PATH}:.
+export NIXOPS_STATE=localstate.nixops
+```
+
+We can now inspect the test network we just created with `nixops
+info`:
+
+```console
+$ nixops info
+Network name: test
+Network UUID: f29084f8-570c-11e5-83af-542696dce997
+Network description: test-network
+Nix expressions: <network.nix>
+
+[...]
+```
+
+And deploy:
+
+```console
+$ nixops deploy
+test-keys> uploading EC2 key pair ‘charon-f29084f8-570c-11e5-83af-542696dce997-test-keys’...
+server...> creating EC2 instance (AMI ‘ami-0126a576’, type ‘t1.micro’, region ‘eu-west-1’)...
+server...> waiting for spot instance request ‘sir-01pq1vda’ to be fulfilled... [pending-evaluation] [fulfilled]
+server...> cancelling spot instance request ‘sir-01pq1vda’... [request-canceled-and-instance-running]
+server...> waiting for IP address... [pending] [running] 54.76.8.50 / 172.31.30.116
+server...> waiting for SSH......
+```
+
+
+### Trick 4: Sharing secrets in git
+
+Keeping AWS keys or any secrets in git unencrypted is a recipe for
+disaster. Bitcoin miners constantly scan Github for accidental AWS key
+uploads and will have spent thousands of your pounds before you can
+blink. The NixOps state file itself contains SSH root keys, exposing
+those keys is the same as having the machine owned with a zero day exploit.
+
+**Keeping either the statefile or the `.envrc` with your AWS keys
+in git unencrypted is grossly negligent.** I know this is common
+knowledge but I prefer to over-communicate on this issue.
+
+We're using a tool called
+[git-crypt](https://github.com/AGWA/git-crypt) to securely share
+secrets in a git repository. git-crypt creatively uses the
+`.gitattribute` file to encrypt certain files in the repository.
+
+First we initialize git-crypt:
+
+```console
+$ git-crypt init
+Generating key...
+$ git-crypt status
+not encrypted: .envrc
+not encrypted: localstate.nixops
+not encrypted: network.nix
+```
+
+Now we set up the `.gitattributes` file:
+
+```shell
+# .gitattributes
+localstate.nixops filter=git-crypt
+.envrc filter=git-crypt diff=git-crypt
+```
+
+Double-check the encryption works:
+
+```console
+$ git-crypt status
+    encrypted: .envrc
+not encrypted: .gitattributes
+    encrypted: localstate.nixops *** WARNING: diff=git-crypt attribute not set ***
+not encrypted: network.nix
+```
+
+Note that we didn't set `diff=git-crypt` on `localstate.nixops`
+because it is a binary SQLite file.
+
+
+### A safer .gitattributes
+
+Keeping encrypted files in many different places in the repository can
+cause accidental exposure due to human mistakes so we generally have a
+`secrets` directory for all the secret things (except for `.envrc`
+which needs to stay top-level).
+
+For example:
+
+```shell
+# .gitattributes
+secrets/* filter=git-crypt
+.envrc filter=git-crypt diff=git-crypt
+```
+
+### Pinning a specific nixos release
```

Comment:
All people with this repo must use the same base reference - this is akin to e.g. setting the base image in docker to `ubuntu 14.04`. I rephrased a bit to make it clearer but maybe I need a different headline?